### PR TITLE
[WiP] Introduce BaseExchange class

### LINKED ISSUE #2367: List-timeframes validates whitelist and load markets
The `list-timeframes` command, when used with `--exchange` command (i.e. when used for exchange which is different from one, specified in the config) performs all the checks, that are needed for trading (validates whitelist, for example), loads markets (which also can fail for some exchanges) -- does many things that are needed for trading, but is redundant for the utility subcommands.

```
$ cp config_binance.json.example config.json
$ freqtrade list-timeframes --exchange bittrex
2019-10-12 11:38:23,040 - freqtrade.loggers - INFO - Verbosity set to 0
2019-10-12 11:38:23,041 - freqtrade.configuration.configuration - INFO - Dry run is enabled
2019-10-12 11:38:23,041 - freqtrade.configuration.configuration - INFO - Using DB: "sqlite://"
2019-10-12 11:38:23,041 - freqtrade.configuration.configuration - INFO - Using max_open_trades: 3 ...
2019-10-12 11:38:23,041 - freqtrade.configuration.configuration - INFO - Using exchange bittrex
2019-10-12 11:38:23,042 - freqtrade.configuration.configuration - INFO - Using user-data directory: /home/user/freqtrade-wrk/github-hroff-1902/freqtrade/user_data ...
2019-10-12 11:38:23,042 - freqtrade.configuration.configuration - INFO - Using data directory: /home/user/freqtrade-wrk/github-hroff-1902/freqtrade/user_data/data/bittrex ...
2019-10-12 11:38:23,042 - freqtrade.configuration.check_exchange - INFO - Checking exchange...
2019-10-12 11:38:23,043 - freqtrade.configuration.check_exchange - INFO - Exchange "bittrex" is officially supported by the Freqtrade development team.
2019-10-12 11:38:23,043 - freqtrade.configuration.configuration - INFO - Using pairlist from configuration.
2019-10-12 11:38:23,043 - freqtrade.resolvers.exchange_resolver - INFO - No Bittrex specific subclass found. Using the generic class instead.
2019-10-12 11:38:23,043 - freqtrade.exchange.exchange - INFO - Instance is running with dry_run enabled
2019-10-12 11:38:23,043 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True}
2019-10-12 11:38:23,046 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True, 'rateLimit': 200}
2019-10-12 11:38:23,060 - freqtrade.exchange.exchange - INFO - Using Exchange "Bittrex"
2019-10-12 11:38:25,255 - freqtrade - ERROR - Pair AST/BTC is not available on Bittrex. Please remove AST/BTC from your whitelist.
```

It should initialize and use a simplified variant of exchange object.

### PR DESCRIPTION
**Pull request by hroff-1902**
Resolves #2367 

```
$ cp config_binance.json.example config.json
$ freqtrade list-timeframes --exchange gdax
2019-10-12 11:50:35,943 - freqtrade.loggers - INFO - Verbosity set to 0
2019-10-12 11:50:35,944 - freqtrade.configuration.configuration - INFO - Dry run is enabled
2019-10-12 11:50:35,944 - freqtrade.configuration.configuration - INFO - Using DB: "sqlite://"
2019-10-12 11:50:35,944 - freqtrade.configuration.configuration - INFO - Using max_open_trades: 3 ...
2019-10-12 11:50:35,944 - freqtrade.configuration.configuration - INFO - Using exchange gdax
2019-10-12 11:50:35,944 - freqtrade.configuration.configuration - INFO - Using user-data directory: /home/user/freqtrade-wrk/github-hroff-1902/freqtrade/user_data ...
2019-10-12 11:50:35,945 - freqtrade.configuration.configuration - INFO - Using data directory: /home/user/freqtrade-wrk/github-hroff-1902/freqtrade/user_data/data/gdax ...
2019-10-12 11:50:35,945 - freqtrade.configuration.check_exchange - INFO - Checking exchange...
2019-10-12 11:50:35,945 - freqtrade.configuration.check_exchange - WARNING - Exchange "gdax" is known to the the ccxt library, available for the bot, but not officially supported by the Freqtrade development team. It may work flawlessly (please report back) or have serious issues. Use it at your own discretion.
2019-10-12 11:50:35,945 - freqtrade.configuration.configuration - INFO - Using pairlist from configuration.
2019-10-12 11:50:35,945 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True}
2019-10-12 11:50:35,948 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True, 'rateLimit': 200}
2019-10-12 11:50:35,960 - freqtrade.exchange.exchange - INFO - Using Exchange "GDAX"
Timeframes available for the exchange `gdax`: 1m, 5m, 15m, 1h, 6h, 1d
```

TODO:
* [x] Fix exchange patching in conftest
* [x] Fix exchange tests
* [x] Add tests for BaseExchange

### CODE REVIEW

**Comment by hroff-1902** [comment]
wtf? (Travis build)

File: freqtrade/exchange/exchange.py [hunk 1]
```diff
@@ -268,10 +218,107 @@ def id(self) -> str:
...
+    _ft_has: Dict = {}
+
+    def __init__(self, config: dict) -> None:
+        """
+        Initializes this module with the given config,
+        it does basic validation whether the specified exchange and pairs are valid.
+        :return: None
+        """
+
+        BaseExchange.__init__(self, config)
```
Comment (xmatthias) [inline freqtrade/exchange/exchange.py]: i think  this should be `super().__init()` making the parent class name irrelevant... (i know this version works too)

File: freqtrade/exchange/exchange.py [hunk 2]
```diff
@@ -139,21 +139,9 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
-class Exchange:
+class BaseExchange:
```
Comment (xmatthias) [inline freqtrade/exchange/exchange.py]: I think this should move to it's own file - that would make the code easier to read and refind.
(i hate to scroll to line 200 to find the class-definition ...)

It's fine to have small classes in one file i guess - but this file is over 1000 lines long - so every line removed will help :D

**Comment by xmatthias** [comment]
I'll close this PR as i think this will create more problems than what it solves.

You can instanciate `BaseExchange` - but you cannot instanciate a Binance class that only inherits from `BaseExchange` - so we would need to have a `BinanceBase` class (and that obviously for all exchanges that have their own class.
Now for Binance this may not be a problem, but other classes may need to pass an extra parameter to load-markets (or something similar, very generic) - making this approach very problematic.

**Pull request closed without merging**